qiskit/algorithms/optimizers/steppable_optimizer.py

```diff
@@ -0,0 +1,213 @@
+    The optimizer then updates its state accordingly and the user can decide whether to stop the optimization process or to repeat a step.
+
+    This interface is more customizable, and allows the user to have full control over the evaluation of the function.
+
+    For example:
+    .. code-block::python
+
+        def objective(x):
+            if random.choice([True, False]):
+                return None
+            else:
+                return (np.linalg.norm(x) - 1) ** 2
+
+        def grad(x):
+            if random.choice([True, False]):
+                return None
+            else:
+                return 2 * (np.linalg.norm(x) - 1) * x / np.linalg.norm(x)
+
+
+        initial_point = np.random.normal(0, 1, size=(100,))
+
+        optimizer = SteppableGradientDescent(maxiter=20)
+        optimizer.initialize(x0=initial_point, fun=objective, jac=grad)
+
+        for _ in range(20):
+            ask_object = optimizer.ask()
+            evaluated_gradient = None
+
+            while evaluated_gradient is None:
+                evaluated_gradient = grad(ask_object.x_center)
+                optimizer._state.njev += 1
+
+            tell_object = GD_TellObject(gradient=evaluated_gradient)
+            optimizer.tell(ask_object=ask_object, tell_object=tell_object)
+
+        result = optimizer.create_result()
+
+    In this case the evaluation of the function has a chance of failing. The user, with specific knowledge about his function can catch this errors and handle before
+    passing the result to the optimizer.
+
+    In case the user isn't dealing with complicated function and is more familiar with step by step optimization algorithms, a method step() has been created that
+    acts as a wrapper for ask() and tell().
+    In the same spirit the method minimize() will optimize the function and return the result without the user having to worry about the optimization process.
+
+    To see other libraries that use this interface one can visit: https://optuna.readthedocs.io/en/stable/tutorial/20_recipes/009_ask_and_tell.html
+
+
+    """
+
+    def __init__(
+        self,
+        callback: Optional[CALLBACK] = None,
+        maxiter: int = None,
+
+    ):
+        """
+        Args:
+            maxiter: Number of steps in the optimization process before ending the loop.
+        """
+        super().__init__()
+        self._state = None
+        self.callback = callback
+        self.maxiter = maxiter  # Remove maxiter
+
+    def ask(self) -> AskObject:
+        """
+        Returns the information on how to evaluate the objective function.
+        Canonical optimization workflow using ask() and tell() can be seen in SteppableOptimizer.step().
+        Returns:
+            AskOjbect: Since the way to evaluate the function can vary much with different optimization algorithms, the object will be
+        a custom dataclass for each optimizer.
```

Comment:
Since we already have the type documented in the signature, we don't need it here anymore I think 🙂 
```suggestion

        Returns:
            Since the way to evaluate the function can vary much with different 
            optimization algorithms, the object will be a custom dataclass for each optimizer.
```